### PR TITLE
Implement Apache DBCP2 datasource metrics

### PR DESCRIPTION
**Pull request by agoallikmaa**

### CODE REVIEW

File: instrumentation/apache-dbcp-2.0/library/src/main/java/io/opentelemetry/instrumentation/apachedbcp/DataSourceMetrics.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.metrics.ObservableLongUpDownCounter;
+import io.opentelemetry.instrumentation.api.metrics.db.DbConnectionPoolMetrics;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.commons.dbcp2.BasicDataSourceMXBean;
+
+public final class DataSourceMetrics {
```
Comment (trask) [inline instrumentation/apache-dbcp-2.0/library/src/main/java/io/opentelemetry/instrumentation/apachedbcp/DataSourceMetrics.java:17]: can this be package-private and accessed only through ApacheDbcpTelemetry?

File: instrumentation/apache-dbcp-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/BasicDataSourceInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      }
+      telemetry().registerMetrics(dataSource, dataSourceName);
+    }
+  }
+
+  @SuppressWarnings("unused")
+  public static class PostDeregisterAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void onExit(@Advice.This BasicDataSource dataSource) {
+      DataSourceMetrics.unregisterMetrics(dataSource);
```
Comment (trask) [inline instrumentation/apache-dbcp-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/BasicDataSourceInstrumentation.java:62]:
```suggestion
      telemetry().unregisterMetrics(dataSource);
```

File: instrumentation/apache-dbcp-2.0/library/src/main/java/io/opentelemetry/instrumentation/apachedbcp/ApacheDbcpTelemetry.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.apachedbcp;
+
+import io.opentelemetry.api.OpenTelemetry;
+import org.apache.commons.dbcp2.BasicDataSourceMXBean;
+
+/** Entrypoint for instrumenting Oracle UCP database connection pools. */
```
Comment (trask) [inline instrumentation/apache-dbcp-2.0/library/src/main/java/io/opentelemetry/instrumentation/apachedbcp/ApacheDbcpTelemetry.java:11]:
```suggestion
/** Entrypoint for instrumenting Apache DBCP database connection pools. */
```

File: instrumentation/apache-dbcp-2.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/ApacheDbcpInstrumentationTest.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  @Override
+  protected InstrumentationExtension testing() {
+    return testing;
+  }
+
+  @Override
+  protected void configure(BasicDataSource dataSource, String dataSourceName) throws Exception {
+    dataSource.preRegister(
+        ManagementFactory.getPlatformMBeanServer(),
+        new ObjectName("io.opentelemetry.db:name=" + dataSourceName));
```
Comment (trask) [inline instrumentation/apache-dbcp-2.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/ApacheDbcpInstrumentationTest.java:30]: we should probably document somewhere that capturing Apache DBCP metrics via javaagent requires the user to add this code

**Review by trask — APPROVED** [review]

File: instrumentation/apache-dbcp-2.0/library/src/main/java/io/opentelemetry/instrumentation/apachedbcp/ApacheDbcpTelemetry.java (hunk 3 above)
Comment (agoallikmaa) [reply instrumentation/apache-dbcp-2.0/library/src/main/java/io/opentelemetry/instrumentation/apachedbcp/ApacheDbcpTelemetry.java]: Fixed.

File: instrumentation/apache-dbcp-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/BasicDataSourceInstrumentation.java (hunk 2 above)
Comment (agoallikmaa) [reply instrumentation/apache-dbcp-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/BasicDataSourceInstrumentation.java]: Fixed.

File: instrumentation/apache-dbcp-2.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/ApacheDbcpInstrumentationTest.java (hunk 4 above)
Comment (agoallikmaa) [reply instrumentation/apache-dbcp-2.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachedbcp/ApacheDbcpInstrumentationTest.java:30]: Added a readme to the `javaagent` module explaining this. However, the auto-instrumentation is intended for use cases where all beans get automatically registered anyway, such as Spring Boot.

**Pull request merged**